lib/bosh/BoshDirectorClient.js

```diff
@@ -16,48 +16,139 @@ const NotFound = errors.NotFound;
+    this.oobDirectorConfigs = BoshDirectorClient.getOobDirectorConfigs();
+    this.populateCache();
+  }
+
+  clearCache() {
+    this.cache = {};
+  }
+
+  static getInfrastructure() {
+    return BoshDirectorClient.getActivePrimary()[0].infrastructure;
+  }
+
+  static getPrimaryConfigs() {
+    return _.filter(config.directors, function (director) {
+      return director.primary;
+    });
+  }
+
+  static getActivePrimary() {
+    return _.filter(config.directors, function (director) {
+      return director.primary && director.support_create;
+    });
+  }
+
+  static getOobDirectorConfigs() {
+    return _.filter(config.directors, function (director) {
+      return !director.primary;
+    });
+  }
+
+  getConfigByName(name) {
+    return _.head(_.filter(config.directors, (director) => director.name === name));
+  }
+
+  populateCache() {
+    this.clearCache();
+    return Promise
+      .map(config.directors, directorConfig => {
+        return this
+          .getDeploymentsByConfig(directorConfig)
+          .tap(deployments => _.map(deployments, deployment => this.cache[deployment.name] = directorConfig));
+      })
+      .tap(() => logger.debug('Cached Deployments:', this.cache));
+  }
+
+  deleteCacheEntry(deploymentName) {
+    return delete this.cache[deploymentName];
+  }
+
+  getDirectorConfig(deploymentName) {
+    return Promise.try(() => {
+      logger.info(`Finding the correct director config for:`, deploymentName);
+      const cache_val = this.cache[deploymentName];
+      if (cache_val !== undefined) {
+        logger.debug('found director in cache...', cache_val);
+        return cache_val;
+      }
+      logger.debug('cache miss for', deploymentName);
+      return this
+        .populateCache()
+        .then(() => {
+          const cache_val = this.cache[deploymentName];
+          if (cache_val !== undefined) {
+            return cache_val;
+          } else {
+            throw new errors.NotFound(`Deployment not found in directors`, deploymentName);
+          }
+        });
     });
-    this.uuid = boshDirector.uuid;
-    this.cpi = boshDirector.cpi;
+  }
+
+  makeRequest(requestDetails, expectedStatusCode, deploymentName) {
+    return this.getDirectorConfig(deploymentName)
+      .then(directorConfig => this.makeRequestWithConfig(requestDetails, expectedStatusCode, directorConfig));
+  }
+
+  makeRequestWithConfig(requestDetails, expectedStatusCode, directorConfig) {
+    requestDetails.baseUrl = directorConfig.url;
+    requestDetails.auth = {
+      user: directorConfig.username,
+      pass: directorConfig.password
+    };
+    requestDetails.rejectUnauthorized = !directorConfig.skip_ssl_validation;
+    return this.request(requestDetails, expectedStatusCode);
   }
 
   getInfo() {
     return this
-      .request({
+      .makeRequestWithConfig({
         method: 'GET',
         url: '/info'
-      }, 200)
+      }, 200, _.sample(this.activePrimary))
       .then(res => JSON.parse(res.body));
```

Comment:
Ok. But later down the line lets add a function to return info of all directors.